### PR TITLE
Support of read only mount of filesystem volumes added

### LINKED ISSUE #2402: Add support of Ceph-CSI shallow volumes
We have request from a customer to support [shallow volumes of Ceph-CSI](https://github.com/ceph/ceph-csi/blob/devel/docs/cephfs-snapshot-backed-volumes.md). Snapshot-backed volumes allow CephFS subvolume snapshots to be exposed as regular read-only PVCs. No data cloning is performed and provisioning such volumes is done in constant time.

For more details please refer to [Snapshots as shallow read-only volumes](https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md) design document.

Since our product uses Kanister, we would like to have  support of read-only mount of filesystem volumes in Kanister. Then we will able to satisfy of customer's requirement.

### PR DESCRIPTION
**Pull request by k0taperk0t**
## Change Overview
This PR adds support of read-only mount required by [Ceph-CSI shallow volumes feature](https://github.com/ceph/ceph-csi/blob/devel/docs/cephfs-snapshot-backed-volumes.md), only for filesystem volumes.

PR series:
* https://github.com/kanisterio/kanister/pull/2376
* https://github.com/kanisterio/kanister/pull/2366
* https://github.com/kanisterio/kanister/pull/2378
* https://github.com/kanisterio/kanister/pull/2377
* https://github.com/kanisterio/kanister/pull/2379
* https://github.com/kanisterio/kanister/pull/2383
* https://github.com/kanisterio/kanister/pull/2386
* => https://github.com/kanisterio/kanister/pull/2365

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #2402 

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

**Comment by k0taperk0t** [comment]
I agree to the DCO for all the commits in this PR.

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,4 @@
 /bin
 /dist
 **/*.swp
+/.idea
```
Comment (pavannd1) [inline .gitignore:10]: Is this a standard directory? Why is it required?
```suggestion
/.idea

```

File: pkg/kube/job.go [hunk 2]
```diff
@@ -73,7 +73,8 @@ func NewJob(clientset kubernetes.Interface, jobName string, namespace string, se
 // Create creates the Job in Kubernetes.
 func (job *Job) Create() error {
```
Comment (pavannd1) [inline pkg/kube/job.go:74]: Should the caller pass in the parent context through the `Create` call?

File: pkg/kube/job.go [hunk 3]
```diff
@@ -125,21 +126,37 @@ func (job *Job) Create() error {
 	return nil
 }
 
-func createFilesystemModeVolumeSpecs(vols map[string]string) (volumeMounts []v1.VolumeMount, podVolumes []v1.Volume, error error) {
+func createFilesystemModeVolumeSpecs(ctx context.Context, cli kubernetes.Interface, namespace string, vols map[string]string) (volumeMounts []v1.VolumeMount, podVolumes []v1.Volume, error error) {
```
Comment (pavannd1) [inline pkg/kube/job.go:129]:
```suggestion
func createFilesystemModeVolumeSpecs(
	ctx context.Context,
	cli kubernetes.Interface,
	namespace string,
	vols map[string]string,
) (volumeMounts []v1.VolumeMount, podVolumes []v1.Volume, error error) {
```

File: pkg/kube/pod.go [hunk 4]
```diff
@@ -207,11 +207,13 @@ func ContainerNameFromPodOptsOrDefault(po *PodOptions) string {
 
 // CreatePod creates a pod with a single container based on the specified image
 func CreatePod(ctx context.Context, cli kubernetes.Interface, opts *PodOptions) (*v1.Pod, error) {
-	pod, err := GetPodObjectFromPodOptions(cli, opts)
+	pod, err := GetPodObjectFromPodOptions(ctx, cli, opts)
 	if err != nil {
 		return nil, errors.Wrapf(err, "Failed to get pod from podOptions. Namespace: %s, NameFmt: %s", opts.Namespace, opts.GenerateName)
 	}
 
+	log.Debug().WithContext(ctx).Print("Creating POD", field.M{"podSpec": pod.Spec})
```
Comment (pavannd1) [inline pkg/kube/pod.go:215]: Is this for debugging? We should not log the pod spec. There are pods created with sensitive information.
```suggestion
	log.Debug().WithContext(ctx).Print("Creating POD", field.M{"name": pod.Name, "namespace": pod.Namespace})
```

File: .gitignore (hunk 1 above)
Comment (k0taperk0t) [reply .gitignore:10]: It's directory for GoLand settings, which is pretty popular IDE for GoLang, so I thing this change might be useful for many people

File: pkg/kube/job.go (hunk 2 above)
Comment (k0taperk0t) [reply pkg/kube/job.go:74]: I tried not to change public interface to decrease painful from my change to other developers. But if you recommend to pass context via `Create` method, I'll could implement this.

File: .gitignore (hunk 1 above)
Comment (k0taperk0t) [reply .gitignore:10]: moved to another PR: https://github.com/kanisterio/kanister/pull/2376

File: pkg/kube/pod.go (hunk 4 above)
Comment (k0taperk0t) [reply pkg/kube/pod.go]: fixed

**Comment by k0taperk0t** [comment]
These changes needed for [K10-18331 Support CephFS snapshot-backed (shallow) volumes](https://kasten.atlassian.net/browse/K10-18331)
TL;DR; To support [Ceph-CSI shallow volumes feature](https://github.com/ceph/ceph-csi/blob/devel/docs/cephfs-snapshot-backed-volumes.md), we need to mount snapshot (temp PVC, which has snapshot as source) in ReadOnlyMany mode, then using of Ceph-CSI shallow volumes will be possible.

File: pkg/kube/job.go (hunk 2 above)
Comment (pavannd1) [reply pkg/kube/job.go:74]: I saw the discussion with Vivek. He's right about the Jobs no longer being used anywhere. This is okay for now but separately we can work on removing unused code.

File: pkg/controllers/repositoryserver/utils.go [hunk 5]
```diff
@@ -253,8 +253,8 @@ func getCondition(status metav1.ConditionStatus, reason string, message string,
 	}
 }
 
-func getVolumes(ctx context.Context, cli kubernetes.Interface, secret *corev1.Secret, namespace string) (map[string]string, error) {
-	vols := make(map[string]string, 0)
+func getVolumes(ctx context.Context, cli kubernetes.Interface, secret *corev1.Secret, namespace string) (map[string]kube.VolumeMountOptions, error) {
```
Comment (pavannd1) [inline pkg/controllers/repositoryserver/utils.go:256]:
```suggestion
func getVolumes(
	ctx context.Context,
	cli kubernetes.Interface,
	secret *corev1.Secret,
	namespace string,
) (map[string]kube.VolumeMountOptions, error) {
```
Improves readability

File: pkg/function/copy_volume_data.go [hunk 6]
```diff
@@ -69,20 +69,25 @@ func (*copyVolumeDataFunc) Name() string {
 	return CopyVolumeDataFuncName
 }
 
-func copyVolumeData(ctx context.Context, cli kubernetes.Interface, tp param.TemplateParams, namespace, pvc, targetPath, encryptionKey string, podOverride map[string]interface{}) (map[string]interface{}, error) {
+func copyVolumeData(ctx context.Context, cli kubernetes.Interface, tp param.TemplateParams, namespace, pvcName, targetPath, encryptionKey string, podOverride map[string]interface{}) (map[string]interface{}, error) {
```
Comment (pavannd1) [inline pkg/function/copy_volume_data.go:72]:
```suggestion
func copyVolumeData(
	ctx context.Context,
	cli kubernetes.Interface,
	tp param.TemplateParams,
	namespace,
	pvcName,
	targetPath,
	encryptionKey string,
	podOverride map[string]interface{},
) (map[string]interface{}, error) {
```

File: pkg/function/prepare_data.go [hunk 7]
```diff
@@ -85,17 +85,22 @@ func getVolumes(tp param.TemplateParams) (map[string]string, error) {
...
-		if _, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvc, metav1.GetOptions{}); err != nil {
-			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvc)
+	validatedVols := make(map[string]kube.VolumeMountOptions)
+	for pvcName, mountPoint := range vols {
+		pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvcName)
 		}
+
+		validatedVols[pvcName] = kube.VolumeMountOptions{MountPoint: mountPoint, ReadOnly: kube.IsAccessModesOfPVCContainReadOnly(pvc)}
```
Comment (pavannd1) [inline pkg/function/prepare_data.go:95]:
```suggestion
		validatedVols[pvcName] = kube.VolumeMountOptions{
			MountPoint: mountPoint,
			ReadOnly: kube.IsAccessModesOfPVCContainReadOnly(pvc),
		}
```

File: pkg/function/restore_data.go [hunk 8]
```diff
@@ -115,17 +115,22 @@ func validateAndGetOptArgs(args map[string]interface{}, tp param.TemplateParams)
...
-		if _, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvc, metav1.GetOptions{}); err != nil {
-			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvc)
+	validatedVols := make(map[string]kube.VolumeMountOptions)
+	for pvcName, mountPoint := range vols {
+		pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvcName)
 		}
+
+		validatedVols[pvcName] = kube.VolumeMountOptions{MountPoint: mountPoint, ReadOnly: kube.IsAccessModesOfPVCContainReadOnly(pvc)}
```
Comment (pavannd1) [inline pkg/function/restore_data.go:125]:
```suggestion
		validatedVols[pvcName] = kube.VolumeMountOptions{
			MountPoint: mountPoint,
			ReadOnly: kube.IsAccessModesOfPVCContainReadOnly(pvc),
		}
```

File: pkg/function/restore_data_using_kopia_server.go [hunk 9]
```diff
@@ -184,18 +184,23 @@ func restoreDataFromServer(
...
-			return nil, errors.Wrap(err, "Failed to retrieve PVC from namespace: "+namespace+" name: "+pvc)
+	validatedVols := make(map[string]kube.VolumeMountOptions)
+	// Validate volumes
+	for pvcName, mountPoint := range vols {
+		pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvcName)
 		}
+
+		validatedVols[pvcName] = kube.VolumeMountOptions{MountPoint: mountPoint, ReadOnly: kube.IsAccessModesOfPVCContainReadOnly(pvc)}
```
Comment (pavannd1) [inline pkg/function/restore_data_using_kopia_server.go:195]:
```suggestion
		validatedVols[pvcName] = kube.VolumeMountOptions{
			MountPoint: mountPoint,
			ReadOnly: kube.IsAccessModesOfPVCContainReadOnly(pvc),
		}
```

File: pkg/kube/job.go [hunk 10]
```diff
@@ -40,12 +40,12 @@ type Job struct {
...
 	sa        string
 	// vols is a map of PVC->Mount points to add to the job pod spec
-	vols      map[string]string
+	vols      map[string]VolumeMountOptions
 	clientset kubernetes.Interface
 }
 
 // NewJob creates a new Job object.
-func NewJob(clientset kubernetes.Interface, jobName string, namespace string, serviceAccount string, image string, vols map[string]string, command ...string) (*Job, error) {
+func NewJob(clientset kubernetes.Interface, jobName string, namespace string, serviceAccount string, image string, vols map[string]VolumeMountOptions, command ...string) (*Job, error) {
```
Comment (pavannd1) [inline pkg/kube/job.go:48]:
```suggestion
func NewJob(
	clientset kubernetes.Interface,
	jobName,
	namespace,
	serviceAccount,
	image string,
	vols map[string]VolumeMountOptions,
	command ...string,
) (*Job, error) {
```

File: pkg/kube/job.go [hunk 11]
```diff
@@ -125,21 +126,29 @@ func (job *Job) Create() error {
 	return nil
 }
 
-func createFilesystemModeVolumeSpecs(vols map[string]string) (volumeMounts []v1.VolumeMount, podVolumes []v1.Volume, error error) {
+func createFilesystemModeVolumeSpecs(
+	ctx context.Context,
+	vols map[string]VolumeMountOptions) (volumeMounts []v1.VolumeMount, podVolumes []v1.Volume, error error) {
```
Comment (pavannd1) [inline pkg/kube/job.go:131]:
```suggestion
	vols map[string]VolumeMountOptions,
) (volumeMounts []v1.VolumeMount, podVolumes []v1.Volume, error error) {
```

File: pkg/kube/utils.go [hunk 12]
```diff
@@ -160,3 +160,14 @@ func IsNodeReady(node *v1.Node) bool {
 	}
 	return false
 }
+
+// IsAccessModesOfPVCContainReadOnly return true if AccessModes of PVC contain `ReadOnlyMany`
+func IsAccessModesOfPVCContainReadOnly(pvc *v1.PersistentVolumeClaim) bool {
```
Comment (viveksinghggits) [inline pkg/kube/utils.go:165]:
```suggestion
func PVCHasAccessMode(pvc *v1.PersistentVolumeClaim, mode AccessMode) bool {
```

File: pkg/kube/pod.go [hunk 13]
```diff
@@ -48,6 +48,11 @@ const (
 	defaultContainerName   = "container"
 )
 
+type VolumeMountOptions struct {
+	MountPoint string
```
Comment (viveksinghggits) [inline pkg/kube/pod.go:52]:
```suggestion
	MountPath string
```

Comment (viveksinghggits) [reply pkg/kube/pod.go:52]: do we handle block devices as well, using the same struct, i.e., `VolumeMountsOptions`?

File: pkg/controllers/repositoryserver/utils.go [hunk 14]
```diff
@@ -264,10 +264,17 @@ func getVolumes(ctx context.Context, cli kubernetes.Interface, secret *corev1.Se
 			return nil, errors.New("Claim name not set for file store location secret, failed to retrieve PVC")
 		}
 		claimNameString := string(claimName)
+		pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, claimNameString, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to validate if PVC %s:%s exists", namespace, claimName)
+		}
```
Comment (viveksinghggits) [inline pkg/controllers/repositoryserver/utils.go:270]: it looks like we are making the same GET PVC call twice here, is that intentional?

File: pkg/kube/utils.go (hunk 12 above)
Comment (k0taperk0t) [reply pkg/kube/utils.go:165]: https://github.com/kanisterio/kanister/pull/2365#discussion_r1360299573

File: pkg/controllers/repositoryserver/utils.go (hunk 5 above)
Comment (k0taperk0t) [reply pkg/controllers/repositoryserver/utils.go]: fixed

File: pkg/function/copy_volume_data.go (hunk 6 above)
Comment (k0taperk0t) [reply pkg/function/copy_volume_data.go]: fixed

File: pkg/function/prepare_data.go (hunk 7 above)
Comment (k0taperk0t) [reply pkg/function/prepare_data.go]: fixed

File: pkg/function/restore_data.go (hunk 8 above)
Comment (k0taperk0t) [reply pkg/function/restore_data.go]: fixed

File: pkg/function/restore_data_using_kopia_server.go (hunk 9 above)
Comment (k0taperk0t) [reply pkg/function/restore_data_using_kopia_server.go]: fixed

File: pkg/kube/job.go (hunk 10 above)
Comment (k0taperk0t) [reply pkg/kube/job.go]: fixed

File: pkg/kube/job.go (hunk 11 above)
Comment (k0taperk0t) [reply pkg/kube/job.go]: fixed

File: pkg/controllers/repositoryserver/utils.go (hunk 14 above)
Comment (k0taperk0t) [reply pkg/controllers/repositoryserver/utils.go:276]: fixed

File: pkg/kube/pod.go (hunk 13 above)
Comment (k0taperk0t) [reply pkg/kube/pod.go:52]: I see, in K8s for volume mounts uses `MountPath` term, regardless in all POSIX-compatible OSes for naming directory where volume mounted to uses `MountPoint` term. Ok, I'll rename it.

File: pkg/kube/utils.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (viveksinghggits) [reply pkg/kube/utils.go:165]: The suggestion that I provided will be able to do exactly what the current function is doing without any issues, but it can also be used from other places as well. For example if someone wants to test a PVC contains a specific accessMode or not.

[the remark this answers is not in the document]
Comment (k0taperk0t) [reply pkg/kube/utils.go:165]: It breaks [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle, what is no good IMO. Another thing, that by my experience, too flexible functions could lead to bug, which will be very difficult to find. In this particular case, after implementing of your suggestion
```suggestion
func PVCHasAccessMode(pvc *v1.PersistentVolumeClaim, mode AccessMode) bool {
```
will be possible to accidentally pass incorrect access mode to this function, what will break a code actually.
Summarize, I understand your motivation, but I think leaving this function as is makes code more robust and simple. Thus, if you don't mind, I'd prefer don't apply your suggestion here.

[the remark this answers is not in the document]
Comment (viveksinghggits) [reply pkg/kube/utils.go:165]: No problem at all @k0taperk0t, we can keep it as it is.
But just for arguments sake :laughing: , if we think that caller can pass incorrect value as argument to a function in that case any written function would be problematic right?

[the remark this answers is not in the document]
Comment (viveksinghggits) [reply pkg/kube/utils.go:165]: Also, what do you think about the name `PVCContainsReadOnlyAccessMode` instead of  `IsAccessModesOfPVCContainReadOnly`. Totally optional again.

[the remark this answers is not in the document]
Comment (k0taperk0t) [reply pkg/kube/utils.go:165]: @viveksinghggits , you're right, if caller will pass incorrect value to this function it will not problem of function. But unfortunately it will became our problem when this code will doesn't work, since we are maintainers of this code. 
A lot of best practices from most popular OOP languages, such as C++ and C#, aren't using in Go, but at the same time, some things teats as anti-patterns in C++ & C#, is Go recommended way. For example, functions for boolean logic in many languages starts from Is/Has/Should, but it's anti-pattern for Go. So..ok, I'll rename `IsAccessModesOfPVCContainReadOnly` to `PVCContainsReadOnlyAccessMode`.

[the remark this answers is not in the document]
Comment (k0taperk0t) [reply pkg/kube/utils.go]: fixed

File: pkg/kube/pod.go (hunk 13 above)
Comment (k0taperk0t) [reply pkg/kube/pod.go]: fixed

**Comment by k0taperk0t** [comment]
@viveksinghggits please continue review

**Review by viveksinghggits — APPROVED** [review]

File: pkg/function/restore_data.go [hunk 15]
```diff
@@ -115,17 +115,25 @@ func validateAndGetOptArgs(args map[string]interface{}, tp param.TemplateParams)
...
+	validatedVols := make(map[string]kube.VolumeMountOptions)
+	for pvcName, mountPoint := range vols {
+		pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvcName)
+		}
+
+		validatedVols[pvcName] = kube.VolumeMountOptions{
+			MountPath: mountPoint,
+			ReadOnly:  kube.PVCContainsReadOnlyAccessMode(pvc),
```
Comment (e-sumin) [inline pkg/function/restore_data.go:127]: Since this is duplicate of code from, maybe it worth to extract it and reuse ?

File: pkg/function/restore_data_using_kopia_server.go [hunk 16]
```diff
@@ -184,9 +184,17 @@ func restoreDataFromServer(
...
+	// Validate volumes
+	for pvcName, mountPoint := range vols {
+		pvc, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to retrieve PVC. Namespace %s, Name %s", namespace, pvcName)
+		}
+
+		validatedVols[pvcName] = kube.VolumeMountOptions{
+			MountPath: mountPoint,
+			ReadOnly:  kube.PVCContainsReadOnlyAccessMode(pvc),
```
Comment (e-sumin) [inline pkg/function/restore_data_using_kopia_server.go:197]: The same is here.

File: pkg/kube/job.go [hunk 17]
```diff
@@ -40,12 +40,19 @@ type Job struct {
...
 	sa        string
 	// vols is a map of PVC->Mount points to add to the job pod spec
-	vols      map[string]string
+	vols      map[string]VolumeMountOptions
 	clientset kubernetes.Interface
 }
 
 // NewJob creates a new Job object.
-func NewJob(clientset kubernetes.Interface, jobName string, namespace string, serviceAccount string, image string, vols map[string]string, command ...string) (*Job, error) {
+func NewJob(clientset kubernetes.Interface,
```
Comment (e-sumin) [inline pkg/kube/job.go:48]:
```suggestion
func NewJob(
	clientset kubernetes.Interface,
```

File: pkg/function/restore_data.go (hunk 15 above)
Comment (k0taperk0t) [reply pkg/function/restore_data.go:127]: I'm afraid it's too late to push another commits for this review. But I can prepare another PR with refactoring if it's crucial changes.

Comment (e-sumin) [reply pkg/function/restore_data.go:127]: Not crucial, I was too late :D We can refactor this piece of code next time.
